src/test/consensus/LedgerTrie_test.cpp

```diff
@@ -0,0 +1,633 @@
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abcd"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abcd"].id());
+        }
+        // Tie-breaker by id
+        {
+            LedgerTrie<Ledger> t;
+            LedgerHistoryHelper h;
+            t.insert(h["abcd"],2);
+            t.insert(h["abce"],2);
+
+            BEAST_EXPECT(h["abce"].id() > h["abcd"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abce"].id());
+
+            t.insert(h["abcd"]);
+            BEAST_EXPECT(h["abce"].id() > h["abcd"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abcd"].id());
+        }
+
+        // Tie-breaker not needed
+        {
+            LedgerTrie<Ledger> t;
+            LedgerHistoryHelper h;
+            t.insert(h["abc"]);
+            t.insert(h["abcd"]);
+            t.insert(h["abce"],2);
+            // abce only has a margin of 1, but it owns the tie-breaker
+            BEAST_EXPECT(h["abce"].id() > h["abcd"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abce"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abce"].id());
+
+            // Switch support from abce to abcd, tie-breaker now needed
+            t.remove(h["abce"]);
+            t.insert(h["abcd"]);
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abc"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abc"].id());
+        }
+
+        // Single node larger grand child
+        {
+            LedgerTrie<Ledger> t;
+            LedgerHistoryHelper h;
+            t.insert(h["abc"]);
+            t.insert(h["abcd"],2);
+            t.insert(h["abcde"],4);
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abcde"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abcde"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{5}).id == h["abcde"].id());
+        }
+
+        // Too much uncommitted support from competing branches
+        {
+            LedgerTrie<Ledger> t;
+            LedgerHistoryHelper h;
+            t.insert(h["abc"]);
+            t.insert(h["abcde"],2);
+            t.insert(h["abcfg"],2);
+            // 'de' and 'fg' are tied without 'abc' vote
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abc"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abc"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{5}).id == h["abc"].id());
+
+            t.remove(h["abc"]);
+            t.insert(h["abcd"]);
+
+            // 'de' branch has 3 votes to 2, so earlier sequences see it as
+            // preferred
+            BEAST_EXPECT(t.getPreferred(Seq{3}).id == h["abcde"].id());
+            BEAST_EXPECT(t.getPreferred(Seq{4}).id == h["abcde"].id());
+
+            // However, if you validated a ledger with Seq 5, potentially on
+            // a different branch, you do not yet know if they chose abcd
+            // or abcf because of you, so abc remains preferred
```

Comment:
The problem is if you fully validate some ledger L, there is no guarantee that the parent of L had majority support of *all* nodes in the previous round. The issue arises when you have Byzantine nodes and/or multiple UNLs, since you don't know about the nodes not in your UNL that might've relied on your validation to fully validate a ledger that you are no longer committed to. By treating all validators whose last validation is for sequence less than `largestIssued`, you wait to be sure that you can't jump to the wrong branch when a majority of your trusted peers actually did validate your current branch.

https://gist.github.com/bachase/6f92b3907baeb3e8cf3ffd5fd8df5f31 is a motivating example involving Byzantine nodes.